### PR TITLE
[fix] once errorBuilder is called Image widget stops loading images

### LINKED ISSUE #64752: Image widget always crash after a single crash
I am using a flutter Image.file() to show an Image in my app. I am also using the errorBuilder to handle any crashes and show a message to the user. 

I encounter a problem when i do these steps. 
1) Load a good image that works
2) Load a corrupt image into the same Image.File() widget
3) Load the original good image back into the same Image.File() widget

It seems every file change after the corrupt photo is passed in (step 2) will result in the error builder being shown and not the new good image. 

If i don't pass in the corrupt photo in step 2, the photo changes like it should. Is this a bug with the flutter Image() or should I be doing something after it goes into the errorBuilder.

Here is my current setup.

```
Image.file(
        file,
        fit: BoxFit.cover,
        errorBuilder: (BuildContext context, Object exception, StackTrace stackTrace) {
          print("Failed to initialise the file");
          print(stackTrace);
          //  Once an error occurs it always goes in here 
          return Text("an error occurred");
        },
      );

```

The actual error I receive on all file changes when/after the corrupt file is passed in is

```
Could not instantiate image codec.
```

### PR DESCRIPTION
**Pull request by pedromassango**
## Description

Image widget does not show a valid image after loading a bad image. The issue is that the exception thrown by the previous loading is not being reset (set to null) when a new image is loaded successfully.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/64752

## Tests

TBA (to be added)

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x]  I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

**Comment by goderbauer** [comment]
/cc @dnfield 

This PR is also missing tests. Can you add some?

**Comment by pedromassango** [comment]
> /cc @dnfield
> 
> This PR is also missing tests. Can you add some?

Working on that!

File: packages/flutter/lib/src/widgets/image.dart [hunk 1]
```diff
@@ -1185,6 +1185,8 @@ class _ImageState extends State<Image> with WidgetsBindingObserver {
     setState(() {
       _imageInfo = imageInfo;
       _loadingProgress = null;
+      _lastException = null;
+      _lastStack = null;
```
Comment (dnfield) [inline packages/flutter/lib/src/widgets/image.dart:1189]: Do we need to do this in `_handleImageChunk` as well?

Comment (dnfield) [reply packages/flutter/lib/src/widgets/image.dart:1189]: Yes - we should add the same after line 1198.

**Comment by dnfield** [comment]
Change seems fine, let me know if you'd like any help writing a test.

File: packages/flutter/lib/src/widgets/image.dart (hunk 1 above)
Comment (pedromassango) [reply packages/flutter/lib/src/widgets/image.dart:1189]: Yes we do.

**Comment by pedromassango** [comment]
> Change seems fine, let me know if you'd like any help writing a test.

I definitely need some help. Can't find a way of testing this!

**Comment by dnfield** [comment]
@pedromassango - if you take a look at the tests in `test/widgets/image_test.dart`, you can see how we create fake ImageStreams and ImageStreamCompleters.

You should be able to mimic one of those tests and basically just drive a fake ImageStreamCompleter such that you call `reportError` and then afterwards call `setImage` and then pump the tester and make sure the right widget(s) are displaying.

**Comment by pedromassango** [comment]
> @pedromassango - if you take a look at the tests in `test/widgets/image_test.dart`, you can see how we create fake ImageStreams and ImageStreamCompleters.
> 
> You should be able to mimic one of those tests and basically just drive a fake ImageStreamCompleter such that you call `reportError` and then afterwards call `setImage` and then pump the tester and make sure the right widget(s) are displaying.

I tried exactly this approach yesterday, but when I call `streamCompleter.reportError()` the test stops and throws the exception provided in reportError(). Need to find a way to send the exception to the widget without throwing and stopping the test.

**Comment by dnfield** [comment]
You'll want to set `FlutterError.onError` to something else, or use `tester.takeException`.

**Comment by pedromassango** [comment]
onError is not working, it always show me this error:
```
'package:flutter_test/src/binding.dart': Failed assertion: line 754 pos 14: '_pendingExceptionDetails != null': A test overrode FlutterError.onError but either failed to return it to its original state, or had unexpected additional errors that it could not handle. Typically, this is caused by using expect() before restoring FlutterError.onError.
dart:core                                      _AssertionError._throwNew
package:flutter_test/src/binding.dart 754:14   TestWidgetsFlutterBinding._runTest.handleUncaughtError
package:flutter_test/src/binding.dart 759:28   TestWidgetsFlutterBinding._runTest.<fn>
dart:async                                     Future._asyncCompleteWithValue.<fn>
package:fake_async/fake_async.dart 193:32      FakeAsync.flushMicrotasks
package:flutter_test/src/binding.dart 1192:19  AutomatedTestWidgetsFlutterBinding.runTest.<fn>

✖ Load a good image after a bad image was loaded
Exited (1)
```

May be similar to https://github.com/flutter/flutter/issues/34499

**Comment by pedromassango** [comment]
I have the test bellow but the issue is that the `errorBuilder` is not being called and it is the only way to verify that the issue was fixed. Tried using `imageProvider.fail()` but it fails the test case.

```dart
  testWidgets('Load a good image after a bad image was loaded', (WidgetTester tester) async {
    final UniqueKey errorKey = UniqueKey();
    Object caughtException;
    final ui.Image image = await tester.runAsync(createTestImage);
    final TestImageStreamCompleter streamCompleter = TestImageStreamCompleter();
    final TestImageProvider imageProvider =
        TestImageProvider(streamCompleter: streamCompleter);

    Widget app() {
      return Image(
        image: imageProvider,
        errorBuilder: (BuildContext context, Object error, StackTrace stackTrace) {
          caughtException = error;
          return SizedBox.expand(key: errorKey);
        },
      );
    }

    await tester.pumpWidget(app());
    expect(find.byKey(errorKey), findsNothing);

    streamCompleter.reportError(exception: 'throw', stack: StackTrace.empty);
    await tester.pump();

    expect(caughtException, isNotNull);
    expect(find.byType(SizedBox), findsOneWidget);

    streamCompleter.setData(
        imageInfo: ImageInfo(
      image: image,
      scale: 1.0,
      debugLabel: 'test.png',
    ));

    await tester.pump();
    expect(caughtException, isNull);
    expect(find.byType(SizedBox), findsNothing);
  });
```

**Comment by dnfield** [comment]
I think you'll want to create your own custom `ImageProvider` fake that throws an exception (say in `load`) during the first pump and then doesn't throw an exception in the second pump.

File: packages/flutter/test/widgets/image_test.dart [hunk 2]
```diff
@@ -1981,3 +2024,44 @@ class FailingImageProvider extends ImageProvider<int> {
...
+  }
+
+  int _loadCallCount = 0;
+
+  @override
+  ImageStreamCompleter load(Object key, DecoderCallback decode) {
+    _loadCallCount += 1;
+    // Fail on first load, succeed on second and fails again on third load.
+    if (_loadCallCount == 1 || _loadCallCount == 3) {
+      throw throws;
```
Comment (pedromassango) [inline packages/flutter/test/widgets/image_test.dart:2052]: @dnfield I can't understand why the test case throw this on second call.

File: packages/flutter/test/widgets/image_test.dart [hunk 3]
```diff
@@ -1981,3 +2024,44 @@ class FailingImageProvider extends ImageProvider<int> {
...
+          image: image,
+          scale: 1.0,
+        ),
+      ),
+    );
+  }
+
+  @override
+  String toString() => '${describeIdentity(this)}()';
+}
```
Comment (dnfield) [inline packages/flutter/test/widgets/image_test.dart:2067]: nit: new line at EOF

File: packages/flutter/test/widgets/image_test.dart (hunk 2 above)
Comment (dnfield) [reply packages/flutter/test/widgets/image_test.dart:2052]: I'm not clear on what you mean.

Comment (dnfield) [reply packages/flutter/test/widgets/image_test.dart:2052]: Ah I see the test now. Perhaps you could instrument this method to print the load count locally and figure out who's calling it?

Comment (pedromassango) [reply packages/flutter/test/widgets/image_test.dart]: Thanks for this. I found out that the Image widget was calling it.

**Comment by pedromassango** [comment]
@dnfield This is ready for another look!

File: packages/flutter/test/widgets/image_test.dart [hunk 4]
```diff
@@ -1981,3 +2027,36 @@ class FailingImageProvider extends ImageProvider<int> {
     );
   }
 }
+
+class TestImageStreamCompleter2 extends ImageStreamCompleter {
```
Comment (dnfield) [inline packages/flutter/test/widgets/image_test.dart:2031]: nit: Can you pick a more descriptive name here? 

Alternatively, can we just merge this class with TestImageStreamCompleter in here?

**Review by dnfield — APPROVED** [review]
LGTM with nit

File: packages/flutter/test/widgets/image_test.dart (hunk 4 above)
Comment (pedromassango) [reply packages/flutter/test/widgets/image_test.dart:2031]: Yes, I think it is possible to merge.

Comment (pedromassango) [reply packages/flutter/test/widgets/image_test.dart]: @dnfield I think it is better now.

**Review by dnfield — APPROVED** [review]
Still LGTM

**Comment by dnfield** [comment]
Web error appears to be a flake. Can you just push a new empty commit to kick CI?

**Comment by dnfield** [comment]
Something int he test isn't quite right - analysis is complaining.

**Comment by pedromassango** [comment]
> Something int he test isn't quite right - analysis is complaining.

Yes, looks like some API has changed

**Pull request merged**